Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+                    useFormat.modifiers = { };
+            } else
+                useFormat.modifiers = { };
+
+            break;
+        }
+
+        if (!useFormat.fourcc)
+            return std::nullopt;
+
+        if (dumpBufferFormatNegotiation) {
+            LOG(Images, " -> selected format, fourcc=0x%2x, n_modifiers=%zu, use modifiers? %i", useFormat.fourcc, useFormat.modifiers.size(), useDRMModifiers);
+
+            for (unsigned modifierIndex = 0; modifierIndex < useFormat.modifiers.size(); ++modifierIndex)
+                LOG(Images, " -> modifier[%2u]=0x%02lx", modifierIndex, useFormat.modifiers[modifierIndex]);
+        }
+
+        return useFormat;
+    };
+
+    auto bufferFormat = negotiateBufferFormat();
+    if (!bufferFormat.has_value()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to negotiate buffer format");
+        return nullptr;
+    }
+
+    auto gbmBuffer = makeUnique<MemoryMappedGPUBuffer>(size);
+    if (!gbmBuffer->allocate(device, bufferFormat.value())) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->map()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to map GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->createEGLImageFromDMABuf()) {
+        LOG_ERROR("[MemoryMappedGPUBuffer] Failed to export GBM buffer as EGLImage");
+        return nullptr;
+    }
+
+    return gbmBuffer;
+}
+
+bool MemoryMappedGPUBuffer::allocate(struct gbm_device* device, const GLDisplay::DMABufFormat& bufferFormat)
+{
+    m_modifier = DRM_FORMAT_MOD_INVALID;
+    if (bufferFormat.modifiers.size() > 0)
+        m_bo = gbm_bo_create_with_modifiers2(device, m_size.width(), m_size.height(), bufferFormat.fourcc, bufferFormat.modifiers.data(), bufferFormat.modifiers.size(), GBM_BO_USE_RENDERING);
+
+    if (m_bo) {
+        m_modifier = gbm_bo_get_modifier(m_bo);
+
+        if (dumpBufferAllocation) {
+            if (m_bo)
+                LOG(Images, " -> buffer allocated with modifiers (gbm_bo reports modifier 0x%02lx)", m_modifier);
+        }
+    } else {
+        m_bo = gbm_bo_create(device, m_size.width(), m_size.height(), bufferFormat.fourcc, GBM_BO_USE_LINEAR);
+        m_modifier = DRM_FORMAT_MOD_INVALID;
+
+        if (dumpBufferAllocation) {
+            if (m_bo)
+                LOG(Images, " -> buffer allocated without modifiers (gbm_bo reports modifier 0x%2lx, m_modifier 0x%2lx)", gbm_bo_get_modifier(m_bo), m_modifier);
+        }
+    }
+
+    if (m_bo) {
+        if (!exportGBMBufferObjectAsDMABuf())
+            return false;
+
+        if (dumpBufferAllocation)
+            LOG(Images, " -> exporting gbm_bo buffer handle %u as dma-buf, got fd %i! succeeded!", gbm_bo_get_handle(m_bo).u32, m_dmaBufFD);
+
+        return true;
+    }
+
```

Comment:
Shouldn't we log an error in this case and the one above, when we fail to export?